assets/app/view/game/corporate_buy_shares.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+require 'view/game/actionable'
+
+module View
+  module Game
+    class CorporateBuyShares < Snabberb::Component
+      include Actionable
+
+      needs :game, store: true
+
+      def render
+        @step = @game.round.active_step
+        @current_actions = @step.current_actions
+        @entity ||= @game.current_entity
+        children = []
+
+        children.concat(render_sources) if @step.current_actions.include?('corporate_buy_shares')
+
+        h('div.margined', children.compact)
+      end
+
+      def render_sources
+        props = {
+          style: {
+            display: 'inline-block',
+            verticalAlign: 'top',
+          },
+        }
+
+        @step.source_list(@entity).map do |source|
+          children = []
+          if source.player?
+            children << h(Player, player: source, game: @game)
+            children << render_player_input(source)
+          elsif source.corporation? && @game.corporation_available?(source)
+            children << h(Corporation, corporation: source)
+            children << render_corporation_input(source)
+          end
+          h(:div, props, children)
+        end.compact
+      end
+
+      def render_player_input(player)
+        return unless @step.current_actions.include?('corporate_buy_shares')
+
+        input = []
+        player.shares.group_by(&:corporation).values.each do |corp_shares|
+          input << render_player_buttons(corp_shares.group_by(&:percent).values.map(&:first))
+        end
+
+        h('div.margined_bottom', { style: { width: '20rem' } }, input.compact)
+      end
+
+      def render_player_buttons(shares)
+        children = []
+
+        shares
+          .select { |share| @step.can_buy?(@entity, share.to_bundle) }
+          .each do |share|
```

Comment:
i prefer to just an each next inteasd of a select each

shares.each do |share|
next if...

Reply:
rather do select map without the chidlren <<

Reply:
I have taken this pattern from lines 107-112 in view/game/buy_sell_shares.rb

Shall I update there, too?

Reply:
yes

Reply:
I'll submit changes to view/game/buy_sell_shares.rb in a separate PR. There's a few there and I don't want to dilute the purpose of this PR.